### PR TITLE
Add basic support for golden image file testing

### PR DESCRIPTION
**Pull request by tvolkert**
* Add a `matchesGoldenFile()` async matcher that will match
  a finder's widget's rasterized image against a golden file.
* Add support for pluggable image comparison backends
* Add a default backend that does simplistic PNG byte
  comparison on locally stored golden files.
* Add support for `flutter test --update-goldens`, which will
  treat the rasterized image bytes produced during the test
  as the new golden bytes and update the golden file accordingly

Still TODO:

* Add support for the `flutter_test_config.dart` test config hook
* Utilize `flutter_test_config.dart` in `packages/flutter/test`
  to install a backend that retrieves golden files from a dedicated
  `flutter/goldens` repo

https://github.com/flutter/flutter/issues/16859

### CODE REVIEW

**Comment by tvolkert** [comment]
This includes https://github.com/flutter/flutter/pull/17085, which should be reviewed first.

File: packages/flutter_test/lib/src/matchers.dart [hunk 1]
```diff
@@ -2,15 +2,21 @@
...
+import 'dart:ui' as ui;
 
+import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter/rendering.dart';
 import 'package:flutter/services.dart';
 import 'package:meta/meta.dart';
-import 'package:test/test.dart';
+import 'package:test/test.dart' hide TypeMatcher;
+import 'package:test/src/frontend/async_matcher.dart'; // ignore: implementation_imports
```
Comment (gspencergoog) [inline packages/flutter_test/lib/src/matchers.dart]: Is this the only way to define an AsyncMatcher subclass? Too bad we have to ignore lint for it to be possible.

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter_test/lib/src/matchers.dart (hunk 1 above)
Comment (tvolkert) [reply packages/flutter_test/lib/src/matchers.dart]: Alas, yes.

**Pull request merged**